### PR TITLE
Allow decimal values in settings using seconds

### LINKED ISSUE #36940: Use floating point values to set time related settings
Currently all time span settings are being parsed as `UInt64` with 2 problems:
* It's silently ignoring decimal values. For example `SET max_execution_time = 1.5` will set the max_execution_time to 1, and what is worse, `0.5` will be set as `0`.
* The smallest values you can set is 1.

In order to allow for more granularity I'd like to change the value to be a Float64 (positive and finite) instead which would fix both instances and allow reducing timeouts (`connect_timeout`, `max_execution_time`, `http_connection_timeout`, etc) under a second.

Although the internal representation would still be a `TimeDiff` (i.e. `Int64`), this would still be a breaking change since not all UInt64 values can be accurately represented as Float64. Nevertheless, I don't expect anyone to be using values over 2^53.

Any thoughts?

### PR DESCRIPTION
**Pull request by Algunenano**
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Changes how settings using `seconds` as type are parsed to support floating point values (for example: `max_execution_time=0.5`). Infinity or NaN values will throw an exception.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

Maintains milliseconds as it was.
Closes https://github.com/ClickHouse/ClickHouse/issues/36940

### CODE REVIEW

**Comment by Algunenano** [comment]
Aparently there are settings with seconds where negative values are ok (and have a meaning) so I'll remove that limitation.

**Comment by Algunenano** [comment]
clang-tidy is complaining (without reason) but it can be workaround.
arm64 tests are failing at `01287_max_execution_speed` which look unrelated since it's not using any `seconds`-unit settings.

**Comment by Algunenano** [comment]
> arm64 tests are failing at 01287_max_execution_speed which look unrelated since it's not using any seconds-unit settings.

Ok, nevermind. That test is setting
```
SET min_execution_speed_bytes = 800000000000, timeout_before_checking_execution_speed = 0.1;
```

The `timeout_before_checking_execution_speed=0.1`, as far as I can see, has never worked because it got truncated to 0. With the changes it respects the 0.1 and, as expected, will not check the execution time until it reaches 0.1 seconds. I guess that's why the test is failing.

I could find that the following tests are setting `timeout_before_checking_execution_speed` to a decimal value:
* 01287_max_execution_speed.sql
* 00156_max_execution_speed_sample_merge.sql
* 01290_max_execution_speed_distributed.sql

I'm going to lower their values to 0 since that's what they were effectively using before.

**Comment by Algunenano** [comment]
Tests:
* Stateless tests flaky check (address, actions): `01287_max_execution_speed` fails because it's too long (under asan)
* Stress tests: They are complaining about backward compatibility checks but I'm not sure what I'm supposed to be seeing / looking for there.

File: src/Core/SettingsFields.cpp [hunk 1]
```diff
@@ -176,34 +179,85 @@ UInt64 SettingFieldMaxThreads::getAuto()
...
     return ::DB::toString(operator UInt64());
 }
 
-template <SettingFieldTimespanUnit unit_>
-void SettingFieldTimespan<unit_>::parseFromString(const String & str)
+template <>
+SettingFieldSeconds::operator Field() const
+{
+    return static_cast<Float64>(value.totalMicroseconds()) / microseconds_per_unit;
+}
```
Comment (yakov-olkhovskiy) [inline src/Core/SettingsFields.cpp:234]: This change can be source of incompatibility for some weird cases, where we rely on integer. But overall shouldn't be an issue, I believe.

File: src/Core/SettingsFields.cpp [hunk 2]
```diff
@@ -176,34 +179,85 @@ UInt64 SettingFieldMaxThreads::getAuto()
...
 {
     *this = stringToNumber<UInt64>(str);
 }
 
 template <SettingFieldTimespanUnit unit_>
 void SettingFieldTimespan<unit_>::writeBinary(WriteBuffer & out) const
 {
+    /// Note that this is unchanged and returns UInt64 for both seconds and milliseconds for
+    /// compatibility reasons as it's only used by the clients or servers older than
+    /// DBMS_MIN_REVISION_WITH_SETTINGS_SERIALIZED_AS_STRINGS
```
Comment (yakov-olkhovskiy) [inline src/Core/SettingsFields.cpp:260]: I think it's better to comment it in terms of lost precision for seconds due to requirements for backward compatibility - we will lose context of this PR in source code - will be not clear unchanged from what.

Comment (Algunenano) [reply src/Core/SettingsFields.cpp:260]: True. I'll improve the comment.

File: src/Core/SettingsFields.cpp (hunk 1 above)
Comment (Algunenano) [reply src/Core/SettingsFields.cpp:234]: If you have any reservation about a specific case let me know and I'll add some extra tests to make sure everything works as expected.

Comment (yakov-olkhovskiy) [reply src/Core/SettingsFields.cpp:234]: frankly can't find anything at least in code... otherwise it will be a proper conversion I think.

**Review by yakov-olkhovskiy — APPROVED** [review]

**Comment by yakov-olkhovskiy** [comment]
@Algunenano 01287_max_execution_speed is failing due to different output
also 02294_floating_point_second_in_settings is failing

**Comment by yakov-olkhovskiy** [comment]
@Algunenano BTW is this force-push made by github itself? I found that force-push often is very messy...

**Comment by Algunenano** [comment]
> BTW is this force-push made by github itself?

No. I did it because the branch was pretty old and rebased instead of merged master/HEAD.

**Comment by yakov-olkhovskiy** [comment]
@Algunenano seems we hit a rare case of CI fluke here:
https://github.com/ClickHouse/ClickHouse/issues/37549
we need a new commit to rerun failed build - could you please try force push again?

**Comment by Algunenano** [comment]
@mergifyio update

**Comment by Algunenano** [comment]
I guess that you need any new commit so I've asked the bot to update the branch. If you actually need a force push let me know and I'll do it tomorrow.

**Comment by yakov-olkhovskiy** [comment]
@mergify update

**Review by yakov-olkhovskiy — APPROVED** [review]

File: tests/queries/0_stateless/02294_decimal_second_errors.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
+SELECT 1 SETTINGS max_execution_time=NaN; -- { serverError 72 }
+SELECT 1 SETTINGS max_execution_time=Infinity; -- { serverError 72 };
+SELECT 1 SETTINGS max_execution_time=-Infinity; -- { serverError 72 };
+
+-- Ok values
+SELECT 1 SETTINGS max_execution_time=-0.5;
+SELECT 1 SETTINGS max_execution_time=0.5;
```
Comment (tavplubix) [inline tests/queries/0_stateless/02294_decimal_second_errors.sql:7]: `DB::Exception: Timeout exceeded: elapsed 0.505827683 seconds, maximum: 0.5.` :( 
https://s3.amazonaws.com/clickhouse-test-reports/39507/fe1e326701b0566881dc7d51793125a911faa07e/stateless_tests__tsan__[2_5].html

Comment (Algunenano) [reply tests/queries/0_stateless/02294_decimal_second_errors.sql:7]: I'll raise it. There are other test already covering 0.something so we don't need to keep a subsecond value here

**Pull request merged**